lib/ecto/changeset.ex

```diff
@@ -427,7 +427,7 @@ defmodule Ecto.Changeset do
   ## Options
 
     * `:empty_values` - a list of values to be considered as empty when casting.
-      All empty values are discarded on cast. Defaults to `[""]`
+      All empty values are replaced with default values on cast if default values are specified. Defaults to `[""]`
```

Comment:
I am not sure this is clearer. Can you please provide an example of what this sentence is meant to clarify? Thank you.

Reply:
### Current behavior without defaults:

```
defmodule User do
  schema "users" do
      field :first_name, :string
      field :last_name, :string
  end   
  
  def changeset(user, attrs) do
      user
      |> cast(attrs, [:first_name, :last_name], empty_values: [""])
  end
end

user = %User{first_name: "some name"} # user has non-empty first_name
User.changeset(user, %{first_name: ""}) 
=> 
#Ecto.Changeset<
  action: nil,
  changes: %{first_name: ""}, # first_name no discarded from changeset and set to ""
  errors: [],
  data: #User<>,
  valid?: true
> 
```

### Current behavior with defaults:

```
defmodule User do
  schema "users" do
      field :first_name, :string, default: "first_name"
      field :last_name, :string
  end   
  
  def changeset(user, attrs) do
      user
      |> cast(attrs, [:first_name, :last_name], empty_values: [""])
  end
end

user = %User{first_name: "some name"} # user has non-empty first_name
User.changeset(user, %{first_name: ""}) 
=> 
#Ecto.Changeset<
  action: nil,
  changes: %{first_name: "first_name"}, # first_name no discarded from changeset and set to default value "first_name"
  errors: [],
  data: #User<>,
  valid?: true
> 
```

### Expected behavior regardless of whether there is default value for field or not:
```
user = %User{first_name: "some name"} # user has non-empty first_name
User.changeset(user, %{first_name: ""}) 
=> 
#Ecto.Changeset<
  action: nil,
  changes: %{}, # no changes, because fields with empty_values where discarded
  errors: [],
  data: #User<>,
  valid?: true
> 
```

I apologize If there is any issue with construction of the phrase, feel free to correct it.

Reply:
Thank you for the write up! I get a different result for your first code snippet though:

```elixir
iex(2)> user = %User{first_name: "some name"} # user has non-empty first_name
%User{
  __meta__: #Ecto.Schema.Metadata<:built, "users">,
  first_name: "some name",
  id: nil,
  last_name: nil
}
iex(3)> User.changeset(user, %{first_name: ""})
#Ecto.Changeset<
  action: nil,
  changes: %{first_name: nil},
  errors: [],
  data: #User<>,
  valid?: true
>
```

At least on Ecto v3.6, the empty value is always discarded and the default value is used instead. So maybe we should say:

```suggestion
      Empty values are always discarded and the default value for the respective key is used instead. Defaults to `[""]`
```

?

Reply:
Oh, you are right about first example, I made an error. 
Maybe I am wrong in interpretation of "discarded" meaning, but for me it sounds like the change wouldn't be passed to the changeset at all. And I'd rather change this word on something more related to "nullifying".